waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+                                                                 :inner-exception e} e))))
+          t-now (t/now)
+          _ (when-not (t/before? t-now not-on-or-after)
+              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                              {:status 400
+                               :saml-assertion-not-on-or-after not-on-or-after
+                               :t-now t-now})))
+          auth-cookie-expiry-date (t/min-date not-on-or-after (t/plus t-now (t/days 1)))
+          auth-cookie-age-in-seconds (-> t-now
+                           (t/interval auth-cookie-expiry-date)
+                           (t/in-seconds))
+          {:keys [authorization/principal authorization/user] :as auth-params-map}
+          (auth/auth-params-map saml-principal)]
+      (auth/handle-request-auth (constantly {:status 303
+                                             :headers {"location" redirect-url}
+                                             :body ""})
+                                request principal auth-params-map password auth-cookie-age-in-seconds))
+    (throw (ex-info "Missing saml-auth-data from SAML authenticated redirect message"
+                    {:status 400}))))
+
+(defn create-request
+  "Return XML elements that represent a SAML 2.0 auth request."
+  [time-issued saml-service-name saml-id acs-url idp-uri]
+  (str
+    (hiccup.page/xml-declaration "UTF-8")
+    (hiccup/html
+      [:samlp:AuthnRequest
+       {:xmlns:samlp "urn:oasis:names:tc:SAML:2.0:protocol"
+        :ID saml-id
+        :Version "2.0"
+        :IssueInstant time-issued
+        :ProtocolBinding "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+        :ProviderName saml-service-name
+        :IsPassive false
+        :Destination idp-uri
+        :AssertionConsumerServiceURL acs-url}
+       [:saml:Issuer
+        {:xmlns:saml "urn:oasis:names:tc:SAML:2.0:assertion"}
+        saml-service-name]])))
+
+(def instant-format (f/formatters :date-time-no-ms))
+
+(defn make-issue-instant
+  "Converts a date-time to a SAML 2.0 time string."
+  []
+  (f/unparse instant-format (t/now)))
+
+(let [opensaml-bootstrapped-atom (atom false)]
+  (defn create-request-factory!
+    "Creates new requests for a particular service, format, and acs-url."
+    [idp-uri saml-service-name acs-url]
+    ;;; Bootstrap opensaml when we create a request factory.
+    (when (compare-and-set! opensaml-bootstrapped-atom false true)
+      (org.opensaml.DefaultBootstrap/bootstrap))
+    #(create-request (make-issue-instant)
+                     saml-service-name
+                     (str "WAITER-" (utils/make-uuid))
+                     acs-url
+                     idp-uri)))
+
+(defrecord SamlAuthenticator [auth-redirect-endpoint idp-cert idp-uri password saml-request-factory]
+  auth/Authenticator
+  (process-callback [this request scheme operation]
+    (if-let [handler (get {"acs" saml-acs-handler
```

Comment:
We can use `cond` and invoke the handler directly.

Reply:
done